cmd/azure.go

```diff
@@ -0,0 +1,333 @@
+type AzureObjects struct {
+	client storage.BlobStorageClient // Azure sdk client
+}
+
+// Convert azure errors to minio object layer errors.
+func azureToObjectError(err error, bucket, object string) error {
+	e, ok := err.(storage.AzureStorageServiceError)
+	if !ok {
+		return err
+	}
+	switch e.StatusCode {
+	case http.StatusNotFound:
+		return ObjectNotFound{bucket, object}
+	}
+	return err
+}
+
+// Inits azure blob storage client and returns AzureObjects.
+func newAzureLayer(account, key string) (ObjectLayer, error) {
+
+	useHTTPS := true
+	c, err := storage.NewClient(account, key, storage.DefaultBaseURL, globalAzureAPIVersion, useHTTPS)
+	if err != nil {
+		return AzureObjects{}, err
+	}
+	client := c.GetBlobService()
+	return &AzureObjects{client}, nil
+}
+
+// Shutdown - Not relevant to Azure.
+func (a AzureObjects) Shutdown() error {
+	return nil
+}
+
+// StorageInfo - Not relevant to Azure.
+func (a AzureObjects) StorageInfo() StorageInfo {
+	return StorageInfo{}
+}
+
+// MakeBucket - Create bucket.
+func (a AzureObjects) MakeBucket(bucket string) error {
+	return traceError(a.client.CreateContainer(bucket, storage.ContainerAccessTypePrivate))
+}
+
+// GetBucketInfo - Get bucket metadata.
+func (a AzureObjects) GetBucketInfo(bucket string) (BucketInfo, error) {
+	// Azure does not have an equivalent call, hence use ListContainers.
+	resp, err := a.client.ListContainers(storage.ListContainersParameters{
+		Prefix: bucket,
+	})
+	if err != nil {
+		return BucketInfo{}, traceError(err)
+	}
+	for _, container := range resp.Containers {
+		if container.Name == bucket {
+			t, e := time.Parse(time.RFC1123, container.Properties.LastModified)
+			if e != nil {
+				continue
+			}
+			return BucketInfo{
+				Name:    bucket,
+				Created: t,
+			}, nil
+		}
+	}
+	return BucketInfo{}, traceError(BucketNotFound{})
+}
+
+// ListBuckets - Use Azure equivalent ListContainers.
+func (a AzureObjects) ListBuckets() (buckets []BucketInfo, err error) {
+	resp, err := a.client.ListContainers(storage.ListContainersParameters{})
+	if err != nil {
+		return nil, traceError(err)
+	}
+	for _, container := range resp.Containers {
+		t, e := time.Parse(time.RFC1123, container.Properties.LastModified)
+		if e != nil {
+			return nil, traceError(e)
+		}
+		buckets = append(buckets, BucketInfo{
+			Name:    container.Name,
+			Created: t,
+		})
+	}
+	return buckets, nil
+}
+
+// DeleteBucket - Use Azure equivalent DeleteContainer.
+func (a AzureObjects) DeleteBucket(bucket string) error {
+	return traceError(a.client.DeleteContainer(bucket))
+}
+
+// ListObjects - Use Azure equivalent ListBlobs.
+func (a AzureObjects) ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	resp, err := a.client.ListBlobs(bucket, storage.ListBlobsParameters{
+		Prefix:     prefix,
+		Marker:     marker,
+		Delimiter:  delimiter,
+		MaxResults: uint(maxKeys),
+	})
+	if err != nil {
+		return result, traceError(err)
```

Comment:
yes, we can do that. i started adding `azureToObjectError(traceError(err), bucket, prefix)` only on need basis. but i will make this change.